File: packages/flutter-sdk/lib/src/connect/apkaya_connect.dart
import '../apkaya_client.dart';
import '../models/models.dart';
import 'secure_session_storage.dart';

/// Mobile wallet connect — WalletConnect v2 (Reown AppKit) + SIWE + email in-app wallet.
class ApkayaConnect {
  ApkayaConnect({
    required this.client,
    required this.chainId,
    SecureSessionStorage? storage,
    this.siweDomain,
    this.siweUri,
    this.siweStatement,
  }) : storage = storage ?? SecureSessionStorage();

  final ApkayaClient client;
  final int chainId;
  final SecureSessionStorage storage;
  final String? siweDomain;
  final String? siweUri;
  final String? siweStatement;

  String? _address;
  String? _sessionToken;
  AuthSessionResult? _inAppSession;

  String? get address => _address;
  String? get sessionToken => _sessionToken;
  bool get isConnected => _address != null;

  /// Restore persisted in-app email session on cold start.
  Future<void> initialize() async {
    _sessionToken = await storage.loadSiweSessionToken();
    final raw = await storage.loadInAppSession();
    if (raw != null) {
      _inAppSession = AuthSessionResult.fromJson(raw);
      _address = _inAppSession!.address;
      _sessionToken ??= _inAppSession!.sessionToken;
    }
  }

  /// Complete SIWE after an external wallet returns address + signature.
  Future<AuthSessionResult> completeSiwe({
    required String walletAddress,
    required String signature,
    required String message,
  }) async {
    final session = await client.auth.siweVerify(message: message, signature: signature);
    _address = session.address;
    _sessionToken = session.sessionToken;
    await storage.saveSiweSessionToken(session.sessionToken);
    return session;
  }

  /// Request SIWE nonce/message for external wallet signing.
  Future<SiweNonceResult> prepareSiwe(String walletAddress) {
    if (siweDomain == null || siweUri == null) {
      throw StateError('siweDomain and siweUri are required for SIWE.');
    }
    return client.auth.siweNonce(
      address: walletAddress,
      chainId: chainId,
      domain: siweDomain!,
      uri: siweUri!,
      statement: siweStatement,
    );
  }

  /// Start email OTP — returns devCode when ENGINE_AUTH_DEV_LOG_OTP=true.
  Future<String?> requestEmailCode(String email) async {
    final result = await client.auth.emailRequestCode(email);
    return result['devCode'] as String?;
  }

  /// Verify email OTP and establish in-app custody wallet session.
  Future<AuthSessionResult> verifyEmailCode(String email, String code) async {
    final session = await client.auth.emailVerifyCode(email: email, code: code);
    _inAppSession = session;
    _address = session.address;
    _sessionToken = session.sessionToken;
    await storage.saveInAppSession(session.toJson());
    await storage.saveSiweSessionToken(session.sessionToken);
    return session;
  }

  /// Sign via in-app custody wallet (email sessions).
  Future<String> signMessage(String message) async {
    final token = _sessionToken;
    if (token == null) throw StateError('Not connected.');
    final result = await client.auth.inAppSignMessage(token, message);
    return result['signature'] as String;
  }

  /// Queue tx via in-app custody wallet.
  Future<TransactionRecord> sendInAppTransaction({
    required String toAddress,
    String? data,
    String? valueWei,
    String? idempotencyKey,
  }) async {
    final token = _sessionToken;
    if (token == null) throw StateError('Not connected.');
    return client.auth.inAppSendTransaction(
      token,
      chainId: chainId,
      toAddress: toAddress,
      data: data,
      valueWei: valueWei,
      idempotencyKey: idempotencyKey,
    );
  }

  Future<void> disconnect() async {
    _address = null;
    _sessionToken = null;
    _inAppSession = null;
    await storage.clearAll();
  }
}

/// WalletConnect v2 via Reown AppKit — initialize in your app widget tree.
///
/// See packages/flutter-sdk/README.md for AppKit setup (project ID, deep links).
/// After AppKit returns a connected address, call [ApkayaConnect.prepareSiwe] +
/// [ApkayaConnect.completeSiwe] to issue an Engine session JWT.
class ApkayaWalletConnect {
  ApkayaWalletConnect({
    required this.projectId,
    required this.appLinkScheme,
  });

  final String projectId;
  final String appLinkScheme;
}

File: packages/unity-sdk/Runtime/Connect/ApkayaConnect.cs
using System;
using System.Threading.Tasks;

namespace Apkaya.UnitySdk.Connect
{
    /// <summary>
    /// External wallet signer — implement with WalletConnect v2 (QR / deep link) or platform SDK.
    /// The SDK does not embed WalletConnect; your game supplies address + signatures.
    /// </summary>
    public interface IApkayaExternalWallet
    {
        Task<string> ConnectAsync();
        Task<string> SignMessageAsync(string message);
        Task DisconnectAsync();
    }

    /// <summary>
    /// SIWE + optional email in-app session against Engine /auth/* — same flow as @apkaya/connect.
    /// </summary>
    public sealed class ApkayaConnect
    {
        private readonly ApkayaClient _client;
        private readonly int _chainId;
        private readonly string _siweDomain;
        private readonly string _siweUri;
        private readonly string _siweStatement;

        public ApkayaConnect(
            ApkayaClient client,
            int chainId,
            string siweDomain,
            string siweUri,
            string siweStatement = null)
        {
            _client = client;
            _chainId = chainId;
            _siweDomain = siweDomain;
            _siweUri = siweUri;
            _siweStatement = siweStatement;
        }

        public string ConnectedAddress { get; private set; }
        public string SessionToken { get; private set; }
        public bool IsConnected => !string.IsNullOrEmpty(ConnectedAddress);

        /// <summary>
        /// Connect external wallet via WalletConnect (or other) then complete SIWE against Engine.
        /// </summary>
        public async Task<AuthSessionResult> ConnectWithExternalWalletAsync(IApkayaExternalWallet wallet)
        {
            var address = await wallet.ConnectAsync();
            var nonce = await _client.SiweNonceAsync(address, _chainId, _siweDomain, _siweUri, _siweStatement);
            var signature = await wallet.SignMessageAsync(nonce.message);
            var session = await _client.SiweVerifyAsync(nonce.message, signature);
            ConnectedAddress = session.address;
            SessionToken = session.sessionToken;
            return session;
        }

        /// <summary>Email OTP in-app custody wallet (Engine creates backend wallet).</summary>
        public async Task<string> RequestEmailCodeAsync(string email)
        {
            var result = await _client.EmailRequestCodeAsync(email);
            return result.TryGetValue("devCode", out var code) ? code?.ToString() : null;
        }

        public async Task<AuthSessionResult> VerifyEmailCodeAsync(string email, string code)
        {
            var session = await _client.EmailVerifyCodeAsync(email, code);
            ConnectedAddress = session.address;
            SessionToken = session.sessionToken;
            return session;
        }

        public void Disconnect()
        {
            ConnectedAddress = null;
            SessionToken = null;
        }
    }

    /// <summary>
    /// Helper for WalletConnect v2 QR / deep-link UX in Unity.
    /// Display <see cref="PairingUri"/> as a QR code; open on mobile via deep link.
    /// Wire <see cref="IApkayaExternalWallet"/> to your WalletConnect Unity integration
    /// (e.g. WalletConnect Unity Modal / WalletConnectSharp).
    /// </summary>
    public sealed class ApkayaWalletConnectSession
    {
        public string PairingUri { get; private set; }
        public string ConnectedAddress { get; private set; }

        public void SetPairingUri(string uri) => PairingUri = uri;
        public void SetConnectedAddress(string address) => ConnectedAddress = address;
        public void Clear()
        {
            PairingUri = null;
            ConnectedAddress = null;
        }
    }
}
